builder.go

```diff
@@ -175,25 +203,56 @@ func containsEntry(metRequires []string, entry BuildPlanEntry) bool {
 	return false
 }
 
-type processMap map[string]launch.Process
+type processMap struct {
+	typeToProcess map[string]launch.Process
+	defaultType   string
+}
+
+func newProcessMap() processMap {
+	return processMap{
+		typeToProcess: make(map[string]launch.Process),
+		defaultType:   "",
+	}
+}
 
-func (m processMap) add(l []launch.Process) {
-	for _, proc := range l {
-		m[proc.Type] = proc
+// This function adds the processes from listToAdd to processMap
+// it sets m.defaultType to the last default process
+// if a non-default process overrides a default process, it returns its type and unset m.defaultType
+func (m *processMap) add(listToAdd []launch.Process) string {
```

Comment:
This function is a bit complex because we want to warn only if the last default process type is being override by a non default process type.

Example A:
default X --> default Y --> non-default Y
There is no default
We'll warn on Y

Example B:
default X --> default Y --> non-default Y --> default Z
Z is the default
We won't warn on Y

Example C:
default Y --> default X --> non-default Y
X is the default
We won't warn on Y

Reply:
I think in example B it's okay to warn when non-default Y replaces default Y, because at that point Y **is** the default. Does that make it any simpler?

Reply:
First, after the changes [here](https://github.com/buildpacks/lifecycle/pull/520/commits/7b24df17bb72f57cfba08d780795cbaba5aaddca), I think that the function is less complicated.

What would you want to do in the following example:
```
default X --> default Y --> non-default Y --> default Z --> default A --> non-default A --> default B
```
Do you think we should warn on both Y and A?
I don't think so but I'm not sure.
If we want to warn on both of them, the function will look like:
```
func (m *processMap) add(listToAdd []launch.Process) string {
	overriddenProcesses := []string
	for _, procToAdd := range listToAdd {
		if procToAdd.Default {
			m.defaultType = procToAdd.Type
		} else if procToAdd.Type == m.defaultType {
			// non-default process overrides a default process
			m.defaultType = ""
                        overriddenProcesses = append(overriddenProcesses, proceToAdd.Type)
		}
		m.typeToProcess[procToAdd.Type] = procToAdd
	}
        warning := fmt.Sprintf("Warning: redefining the following default process type with a process not marked as default: %s\n", strings.Join(overridenProcesses, ", "))
	return warning
}
```
What do you think?